### ISSUE TITLE
Limitations of the method

### ISSUE THREAD

**Issue by fyang064** [Beomi/InfiniTransformer#13]
Just wondering if any limitations of the Infini-attention like inference speed and model performance. Not too much discussions in the paper.

**Comment by Beomi** [comment]
Since infini attention uses segmentation which is mainly focused on reducing memory usage and computational cost into O(N) so if you use very long seq such as seq len = 1M, then you have to wait 5min per batch on 1x H100 gpu.

In this implementation I used for loop to make it work so you'll get exactly N times more time to inference or to train.

e.g., 
segment 2048 & block 2048 ==> 1 segments, just 1 attention computation
segment 2048 & block 32K ==> 16 segments, slower 16 times than block 2048

Model performance check was done at the paper with PPL but we may need more tests on very long sequence based tests.

**Comment by fyang064** [comment]
> Since infini attention uses segmentation which is mainly focused on reducing memory usage and computational cost into O(N) so if you use very long seq such as seq len = 1M, then you have to wait 5min per batch on 1x H100 gpu.
> 
> In this implementation I used for loop to make it work so you'll get exactly N times more time to inference or to train.
> 
> e.g., segment 2048 & block 2048 ==> 1 segments, just 1 attention computation segment 2048 & block 32K ==> 16 segments, slower 16 times than block 2048
> 
> Model performance check was done at the paper with PPL but we may need more tests on very long sequence based tests.

I agree with you that more tests on very long sequence input are needed. It's not unique to think about using the segmentation to break down the long sequence input by for loop, e.g. RingAttention. Just curious about the impact of compressive memory, how does it influence the computation of attention when dealing with the long input?

**Comment by Lazy3valuation** [comment]
> Since infini attention uses segmentation which is mainly focused on reducing memory usage and computational cost into O(N) so if you use very long seq such as seq len = 1M, then you have to wait 5min per batch on 1x H100 gpu.
> 
> In this implementation I used for loop to make it work so you'll get exactly N times more time to inference or to train.
> 
> e.g., segment 2048 & block 2048 ==> 1 segments, just 1 attention computation segment 2048 & block 32K ==> 16 segments, slower 16 times than block 2048
> 
> Model performance check was done at the paper with PPL but we may need more tests on very long sequence based tests.

I also noticed during various tests that it slows down with time even before generating more tokens than segment dimension. For example, in my small 4b quantized model, it starts with 35tk/s, after about 1k token it's at 4tk/s.
